supervision/detection/tools/csv_sink.py

```diff
@@ -0,0 +1,147 @@
+    "x_max",
+    "y_max",
+    "class_id",
+    "confidence",
+    "tracker_id",
+]
+
+
+class CSVSink:
+    """
+    A utility class for saving detection data to a CSV file. This class is designed to
+    efficiently serialize detection objects into a CSV format, allowing for the inclusion of
+    bounding box coordinates and additional attributes like confidence, class ID, and tracker ID.
+
+    The class supports the capability to include custom data alongside the detection fields,
+    providing flexibility for logging various types of information.
+
+    Args:
+        filename (str): The name of the CSV file where the detections will be stored.
+            Defaults to 'output.csv'.
+
+    Example:
+        ```python
+        import numpy as np
+        import supervision as sv
+        from ultralytics import YOLO
+        import time
+
+        model = YOLO("yolov8n.pt")
+        tracker = sv.ByteTrack()
+        box_annotator = sv.BoundingBoxAnnotator()
+        label_annotator = sv.LabelAnnotator()
+        csv_sink = sv.CSVSink(...)
+
+        def callback(frame: np.ndarray, _: int) -> np.ndarray:
+            start_time = time.time()
+            results = model(frame)[0]
+            detections = sv.Detections.from_ultralytics(results)
+            detections = tracker.update_with_detections(detections)
+
+            labels = [
+                f"#{tracker_id} {results.names[class_id]}"
+                for class_id, tracker_id
+                in zip(detections.class_id, detections.tracker_id)
+            ]
+            time_frame = (time.time() - start_time)
+
+            csv_sink.append(detections, custom_data={"processing_time": time_frame})
+
+            annotated_frame = box_annotator.annotate(
+                frame.copy(), detections=detections)
+            return label_annotator.annotate(
+                annotated_frame, detections=detections, labels=labels)
+
+        csv_sink.open()
+        sv.process_video(
+            source_path="people-walking.mp4",
+            target_path="result.mp4",
+            callback=callback
+        )
+        csv_sink.close()
+        ```
+    """  # noqa: E501 // docs
+
+    def __init__(self, filename: str = "output.csv"):
+        self.filename = filename
+        self.file: Optional[open] = None
+        self.writer: Optional[csv.writer] = None
+        self.header_written = False
+        self.fieldnames = []  # To keep track of header names
+
+    def __enter__(self) -> CSVSink:
+        self.open()
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Optional[type],
+        exc_val: Optional[Exception],
+        exc_tb: Optional[Any],
+    ) -> None:
+        self.close()
+
+    def open(self) -> None:
+        self.file = open(self.filename, "w", newline="")
+        self.writer = csv.writer(self.file)
+
+    def close(self) -> None:
+        if self.file:
+            self.file.close()
+
+    @staticmethod
+    def parse_detection_data(
```

Comment:
Hi @SkalskiP, I added a condition that ensures the `class_id`, `confidence`, and `tracker_id` columns are always present in the CSV, even if they are missing for a particular detection (Optionals values), In case any of these columns are missing, the CSV file will still contain the respective column, but it will be empty. 

Please let me know if that is okay?

Reply:
During testing, I discovered that the `mock_detections` function does not support complex data structures similar to the one provided.
```python
data={"metadata": {"sensor_id": 101, "location": "north"}, "tags": ["urgent", "review"]}
``` 

I got the next issue
```python
test/detection/test_csv.py:112: in <module>
    mock_detections(
test/test_utils.py:19: in mock_detections
    return Detections(
<string>:9: in __init__
    ???
supervision/detection/core.py:61: in __post_init__
    validate_detections_fields(
supervision/detection/utils.py:591: in validate_detections_fields
    validate_data(data, n)
supervision/detection/utils.py:568: in validate_data
    raise ValueError(f"Shape of np.ndarray for key '{key}' must be ({n},)")
E   ValueError: Shape of np.ndarray for key 'tags' must be (1,)
```